File: programming/problem_5/problem_5.dart
void main() {
  // 与えられた中心を基準に回文を展開し、左右のインデックスを返す関数
  Map<String, int>? expandAroundCenter({
    required List<String> strList,
    required int leftIndex,
    required int rightIndex,
  }) {
    if (leftIndex < 0 || strList.length <= rightIndex) {
      return null;
    }

    // 左右のインデックスを境界内で展開し、回文である限り続ける
    while (leftIndex >= 0 &&
        rightIndex < strList.length &&
        strList[leftIndex] == strList[rightIndex]) {
      leftIndex--;
      rightIndex++;
    }

    // 最終的な左右のインデックスを返す
    return {
      'leftIndex': leftIndex + 1,
      'rightIndex': rightIndex - 1,
    };
  }

  // 与えられた文字列内で最長の回文部分文字列を見つける関数
  String? findLongestPalindromeSubstring(String str) {
    final strList = str.split('');
    var longestPalindromeSubstring = '';
    // 文字列の各文字とその隣の文字の間を中心に、奇数長と偶数長の回文を探す
    for (var i = 0; i < strList.length; i++) {
      for (var j = 0; j < 2; j++) {
        final indexMap = expandAroundCenter(
          strList: strList,
          leftIndex: i,
          rightIndex: i + j,
        );

        if (indexMap != null) {
          // 回文部分文字列を取得し、最長のものを更新する
          final palindromeSubstring = strList
              .sublist(indexMap['leftIndex']!, indexMap['rightIndex']! + 1)
              .join('');

          if (longestPalindromeSubstring.length < palindromeSubstring.length) {
            longestPalindromeSubstring = palindromeSubstring;
          }
        }
      }
    }
    return longestPalindromeSubstring.isEmpty
        ? null
        : longestPalindromeSubstring;
  }

  print(
    findLongestPalindromeSubstring('abacdcbeebdbefaafbefdbbeabcdedcbaebdcacba'),
  );
}

File: programming/problem_5/problem_5.go
package main

import "fmt"

// expandedAroundCentner 与えられた中心を基準に回文を展開し、左右のインデックスを返す。
func expandedAroundCentner(stList []rune, leftIndex int, rightIndex int) (newLeftIndex int, newRightIndex int,success bool) {
	if leftIndex < 0 || len(stList) <= rightIndex {
		return 0, 0, false
	}

	// 左右のインデックスを境界内で展開し、回文である限り続ける
	for leftIndex >= 0  && rightIndex < len(stList) && stList[leftIndex] == stList[rightIndex]{
		leftIndex--
		rightIndex++
	}

	return leftIndex + 1, rightIndex - 1, true
}

// findLongestPalindromeSubstring 与えられた文字列内で最長の回文部分文字列を返す
func findLongestPalindromeSubstring(str string) *string {
	strList := []rune(str)
	var longestPalindromeSubstring string

	for i := 0; i < 2*len(strList)-1; i++ {
		leftIndex := i / 2
		rightIndex := leftIndex + i%2

		newLeftIndex, newRightIndex, success := expandedAroundCentner(strList, leftIndex, rightIndex)
		if success {
			palindromeSubstring := string(strList[newLeftIndex:newRightIndex+1])
			if len(longestPalindromeSubstring) < len(palindromeSubstring) {
				longestPalindromeSubstring = palindromeSubstring
			}
		}
	}
	if longestPalindromeSubstring == "" {
		return nil
	}
	return &longestPalindromeSubstring
}

func main()  {
	result := findLongestPalindromeSubstring("abacdcbeebdbefaafbefdbbeabcdeedcbaebdcacba")
	if result == nil {
		fmt.Printf("No palindrome substring found.")
	} else {
		fmt.Printf("Longest palindrome substring %s\n", *result)
	}
}
